### ISSUE TITLE
Making editing snappier over remote networks

### ISSUE THREAD

**Issue by amitds1997** [neovim/neovim#24690]
### Problem

Consider the scenario where my Neovim server is running on a remote server and I connect to it through ssh. Any and every operation that happens, including buffer changes, happen on the remote server. This makes the editing experience on the local feel "laggy" for slow networks. Instead, the editing experience on Neovim should be amazing even when the network is lagging. It's fine if the global search is slow or running system commands is slow, but the editing experience should always be premium and fast.

I looked into other editors that support such behavior:
1. [JetBrains client](https://www.jetbrains.com/help/idea/remote-development-overview.html#859dd19f): JetBrains Client has an advanced editor based on the JetBrains Rider’s RD protocol, which makes typing feel instant.
2. [Lapce](https://docs.lapce.dev/development/architecture): Lapce lets the client do the buffer management and communicates the delta changes in the buffer over to the server which does everything else.
3. VS Code: My guess, is that VS Code, also does something similar to support snappy editing over remote connections, although I could not find any documents supporting this.

Another question is how does this fit into our roadmap and vision for Neovim? If this is something we would be interested in, I would be happy to work towards getting it across the finish line.

### Expected behavior

Editing is fast even when network is unstable or disconnected (obviously for the cached buffers)

**Comment by amitds1997** [comment]
Created this because of [this comment](https://github.com/neovim/neovim/issues/21635#issuecomment-1674698983) from @justinmk. The scope of that issue seemed to be to just set up remote connections through a single command.

This issue is to initiate discussion on how we can achieve this behavior. Let me know if this discussion would be more suitable in that issue @zeertzjq. Thanks!

**Comment by fecet** [comment]
Maybe mosh is also relevant, though I don't think it can work with neovim out of the box.

**Comment by amitds1997** [comment]
> Maybe mosh is also relevant, though I don't think it can work with neovim out of the box.

Mosh will definitely help us perform better on slow networks but it would not help us much with the overall situation that every action has to happen at the remote and travel back to the UI client.

If the buffer changes instead happen locally and the remote server just provides all the additional details, LSP, linters, remote commands, etc. I think we can reduce the net data transfer and make things faster. I am obviously overlooking a lot of actual details and challenges that we would have to get through to make this possible.

I do not want to distract the group but this approach might have further advantages:
1. Actual real time collaboration can be a reality with this. I don't think it's possible right now. But that's a non-goal for this issue.
2. Right now, UIs are auto adjusted to the smallest viewframe available. Shifting this responsibility onto the client would help each client draw it's own UI and operate at different scales independently with the same data.

**Comment by amitds1997** [comment]
I have been leveraging Neovim to work on Dev containers and this would make my life way better so I am willing to put a bit of effort into making this possible. I do realize that this may involve some lengthy design discussions before deciding upon how to achieve this. But if I could have some pointers on what would be the right place to begin at, I would happy to do some initial legwork on this or start a discussion on this.

I do see that this is currently unplanned so I do realize that this is not a priority, and I don't see it being immediately prioritized either but if possible, I do not want this to go stale.

**Comment by CGamesPlay** [comment]
My comment in the other thread was the inspiration for this issue, but the comment was flagged in that thread, so I'm reproducing here. Hopefully I can also explain it a bit better after sitting with it for a week.

**Background** Neovim already supports editing files with native-like performance with minimal latency: netrw. However, modern users also want to run language servers, integrated terminals, and other extensions that netrw does not support. Therefore, we want to add these features to neovim, so that users can edit files in on remote machines without also running the full UI on the remote machine.

**Most basic version** I think the best approach to adding this is to build an abstraction around how Neovim interacts with the host system. The abstraction includes filesystem access, process execution, environment variables, and any other OS interfaces that Neovim uses. Once the abstraction is in place, the "local OS" implementation can be swapped with a "network socket" implementation (e.g. a filesystem access would not go to the local OS, but instead would send a request across a network socket to a server which would fulfill the request).

This approach would mean that all UI is rendered locally (no network latency), but Neovim would act like it was running on the remote system (scripts would not be able to tell they weren't). For some examples, in this state `:pwd` shows a path on the remote system, `!` filters run processes on the remote system, etc.

**Ideal version** If we could get this far, it would be a big win for remote editing with neovim, but I don't think it fully gets us to the ideal state, because while most operations should happen on the "remote" side, there are many cases where we do actually want to run operations on the local side. The most obvious is init files: I want my local vimrc, but I want to edit files on a remote system. 

To implement something like this, there should be some way for Neovim to direct each operation to either the "default" location (the local OS most of the time, but sometimes a network socket) or the "local" location (always the local OS). Some examples might be `:local COMMAND` would run a command in the local location, or a Lua function that runs a closure in the local location.

With this new targeting mechanism in place, Neovim could locate runtime path files using the local location, but execute them using the default location. This means that an existing script doesn't need to do anything to support remote editing: it simply thinks that it's being hosted in a neovim running on the remote system. But new scripts can use escape hatches like `:local` to perform rich integrations with the user's local terminal where appropriate (e.g. opening a browser window).

**Open questions:**

- What needs to be covered under this abstraction? The goal of the default location is to effectively trick any user command or script that isn't "location-aware" that it's actually running on the remote system.
- What core parts of Neovim need to be made location-aware? Init scripts and runtime path files are already mentioned. The integrated terminal will likely require special attention.
- Does there need to be a "remote runtime path" feature or some other way of supporting "remote extensions"? I don't think this is critical and I think that a third-party plugin manager could add support for it instead.
- This proposal does not talk about where the magic "network socket" comes from (i.e. installation o the remote host). I think this is better left to third-party scripts, but we should make sure to leave a way for such a script to install and connect to a remote from within Neovim (e.g. "restart with this modified command line").

Hopefully this helps get the discussion going around the ideal architecture for a no-latency, network-enabled editing experience, and what would have to change in Neovim to reach it. I would love to help work on this feature, but realistically it's low on my priority list (when I am remote editing today, I switch to VSCodium).

**Comment by justinmk** [comment]
Have you tried https://github.com/chipsenkbeil/distant.nvim ? What's missing from it?

**Comment by CGamesPlay** [comment]
Distant looks promising (haven't used it personally), but from the design doc the biggest thing it's missing: it requires all plugins to opt in to executing remotely. If nvim exposed some way for distant to make this behavior opt-out rather than opt-in, I think it could be a viable approach.

What I mean by this: if I want to use a language server plugin, it has to be distant-aware and know to execute the language server remotely rather than on the local system. By default, it won't and will crash because none of the source files are locatable.

So to make this a solution that enables a true seamless remote environment, nvim should provide some hooks that distant could use to make other scripts think they were actually running on the remote system, including for processes, environment, and file system. If nvim exposed these and distant implemented them, the entire existing plugin ecosystem would "just work" remotely, I believe.

**Comment by justinmk** [comment]
> if I want to use a language server plugin, it has to be distant-aware and know to execute the language server remotely rather than on the local system.

[vscode remote](https://code.visualstudio.com/docs/remote/remote-overview) runs extensions on the remote system in a remote vscode server. It doesn't have a layer that allows _local_ extensions to transparently operate in a remote context. That's a [tarpit](https://www.bu.edu/csmet/2018/08/30/re-reading-tanenbaums-critique-of-rpc-30-years-later/).

> nvim should provide some hooks that distant could use to make other scripts think they were actually running on the remote system, including for processes, environment, and file system.

Impossible. And outside the scope of the original request here.

# Proposal

I interpret the "expected behavior" as:

> Make insert/visual/normal-mode fast, for the current buffer.

Most latency-sensitive editing operations do not need to run remotely. The obvious exception is auto-completion. Here's a sketch (can be adjusted/updated) of something that could work:

1. Let plugins declare themselves as "workspace", or "ui". VSCode has this [concept](https://code.visualstudio.com/api/advanced-topics/remote-extensions#architecture-and-extension-kinds).
2. Introduce "remote-sync mode": unlike the current `--remote-ui` where _execution_ is owned by the remote, this alternate mode does:
    - _2-way sync_ of remote <=> local buffers (how [ActualVim](https://github.com/lunixbochs/ActualVim) and [vscode-neovim](https://github.com/vscode-neovim/vscode-neovim/) work).
        - Syncs buffer contents, does _not_ attempt to resolve "edit operations" (CRDT-style).
    - Local Nvim presents the current buffer as a scratch buffer in local Nvim.
    - Local Nvim owns editing of the current (scratch) buffer.
    - Local Nvim enables "ui" plugins on the current (scratch) buffer.
    - Everything else (window layout, cmdline-mode, execution of most plugins and scripts, ...) is owned by the remote.
3. Implement this as a _plugin_, just like vscode-neovim.
    - The edges are getting polished away over time, which indicates this is tractable.
    - We have line-of-sight to concrete improvements needed in Nvim ([example](https://github.com/neovim/neovim/issues/19806)).
4. How does LSP completion work in insert-mode?
    - Note: vscode-neovim delegates insert-mode to vscode (the "remote"). But for this feature request, local Nvim needs to own insert-mode (to avoid latency).
    - `vim.lsp` provides a special "ui" variant of `vim.lsp.omnifunc()` which makes a request to the remote `vim.lsp` and feeds the result into local.
    - Other LSP plugins would need to provide a (thin) "ui" variant which calls into remote Nvim as needed.

**Comment by CGamesPlay** [comment]
You're right, trying to proxy local plugins over to the remote is probably a bad idea. Definitely an oversight on my part. It makes sense for a remote-providing plugin to be responsible for installing "ui" or "workspace" plugins separately. I definitely would like to see a solution that, as much as is possible, doesn't require throwing out all of the existing plugin ecosystem.

What would it mean for a plugin to actually be "workspace" or "ui"? Would this be just a flag that enables a plugin to disable itself if not used in the right context?

Per your second point, is there a technical reason why it couldn't be all open buffers? During a transient lag spike (wifi drops), I would love to be able to switch between opened buffers, referencing one file and typing into another. Similarly, is it impossible to do the window layout locally? Users can resize splits and the entire terminal with the mouse, and which would lead to a pretty laggy experience, and completely break Nvim during a transient lag spike.

**Comment by amitds1997** [comment]
The suggested proposal seems great for the current issue. 

When we talk about plugins under the UI/workspace concept, would a status line plugin like [lualine.nvim](https://github.com/nvim-lualine/lualine.nvim) be a UI plugin or a workspace one? The status line is just a UI component but it needs data that is only available at remote (e.g. the LSP attached to current buffer, current warnings, number of search results found, etc). Would it be possible for plugins to dictate which components of the UI will be owned by them and coming over the network? 

If not, then it might make sense for the entire UI to be owned either by remote (the current way) or the local Nvim.

**Comment by justinmk** [comment]
> What would it mean for a plugin to actually be "workspace" or "ui"?

When connected in "remote-sync mode", 

- "workspace" plugins execute in the remote `nvim` server
- "ui" plugins execute in the local `nvim` client. They don't have any idea about the remote server unless they explicity request it (as in the LSP completion example above).

> is there a technical reason why it couldn't be all open buffers? ... is it impossible to do the window layout locally?

No technical reason, just a constraint that limits scope, without being a "one-way door". If the above proposal reaches maturity, then we'll have a better idea of how feasible (and how important) it is to go further.

> would a status line plugin like [lualine.nvim](https://github.com/nvim-lualine/lualine.nvim) be a UI plugin or a workspace one?

It depends on execution and workspace state so it would be a "workspace" plugin

**Comment by tyalie** [comment]
Just a small ping to ask if there have been any further offline discussions or if other progress has been made towards this goal in the meantime.
